ouroboros-network-framework/src/Ouroboros/Network/Subscription/Dns.hs

```diff
@@ -77,32 +97,34 @@ dnsResolve :: forall m s.
      , MonadThrow m
      )
     => Tracer m DnsTrace
-    -> Resolver m
+    -> m a
+    -> (a -> (Resolver m -> m (SubscriptionTarget m Socket.SockAddr)) -> m (SubscriptionTarget m Socket.SockAddr))
     -> StrictTVar m s
     -> BeforeConnect m s Socket.SockAddr
     -> DnsSubscriptionTarget
     -> m (SubscriptionTarget m Socket.SockAddr)
-dnsResolve tracer resolver peerStatesVar beforeConnect (DnsSubscriptionTarget domain _ _) = do
-    ipv6Rsps <- newEmptyTMVarM
-    ipv4Rsps <- newEmptyTMVarM
-    gotIpv6Rsp <- Lazy.newTVarM False
-
-    -- Though the DNS lib does have its own timeouts, these do not work
-    -- on Windows reliably so as a workaround we add an extra layer
-    -- of timeout on the outside.
-    -- TODO: fix upstream dns lib
-    res <- timeout 20 $ do
-             aid_ipv6 <- async $ resolveAAAA gotIpv6Rsp ipv6Rsps
-             aid_ipv4 <- async $ resolveA gotIpv6Rsp ipv4Rsps
-             rd_e <- waitEitherCatch aid_ipv6 aid_ipv4
-             handleResult ipv6Rsps ipv4Rsps rd_e
-    case res of
-      Nothing -> do
-        -- TODO: the thread timedout, we should trace it
-        return (SubscriptionTarget $ pure Nothing)
-      Just st ->
-        return st
-
+dnsResolve tracer getSeed withResolver peerStatesVar beforeConnect (DnsSubscriptionTarget domain _ _) = do
+    rs <- getSeed
+    withResolver rs $ \resolver -> do
+        ipv6Rsps <- newEmptyTMVarM
+        ipv4Rsps <- newEmptyTMVarM
+        gotIpv6Rsp <- Lazy.newTVarM False
+
+        -- Though the DNS lib does have its own timeouts, these do not work
+        -- on Windows reliably so as a workaround we add an extra layer
+        -- of timeout on the outside.
+        -- TODO: fix upstream dns lib
+        res <- timeout 20 $ do
+                 aid_ipv6 <- async $ resolveAAAA resolver gotIpv6Rsp ipv6Rsps
+                 aid_ipv4 <- async $ resolveA resolver gotIpv6Rsp ipv4Rsps
```

Comment:
I noticed a missing comment here: if an `async` exception will hit the main thread here, the two threads will not be killed. Though we should not do this kind of change lightly, since the dns library on Windows is not using our interruptible `Win32-network`.  @karknu could you add a note here (or I can do that if you wish, in between of waiting for things to compile./ run).